### ISSUE TITLE
Pydantics model validator causing weird issues when in classmethod-decorators

### ISSUE THREAD

**Issue by ollz272** [astral-sh/ruff#13617]
We have the following in our ruff config:

```.toml
[tool.ruff.lint.pep8-naming]
classmethod-decorators = ["classmethod", "pydantic.model_validator", "pydantic.field_validator"]
```

This is so we can do the following in our code:
```python
import pydantic
from pydantic import BaseModel

class Model(BaseModel):
    """Some Model."""

    foo: str
    bar: int

    @pydantic.model_validator(mode="before")
    def some_validation(cls, data):
        """Before Validation."""
        if data["foo"] == "hello" and data["bar"] == 1:
            raise ValueError

        return data
```

This works well, however when we set the validator mode to "after", this is an instance method rather than a class method, so we get:
```python
import pydantic
from pydantic import BaseModel


class Model(BaseModel):
    """Some Model."""

    foo: str
    bar: int

    @pydantic.model_validator(mode="before")
    def before_validation(cls, data):
        """Before Validation."""
        if data["foo"] == "hello" and data["bar"] == 1:
            raise ValueError

        return data

    @pydantic.model_validator(mode="after")
    def after_validation(self):
        """After Validation."""
        if not self.bar:
            raise ValueError

        return self

```

We get the following:
```
 N804 First argument of a class method should be named `cls`
   |
19 |     @pydantic.model_validator(mode="after")
20 |     def after_validation(self):
   |                          ^^^^ N804
21 |         """After Validation."""
22 |         if not self.bar:
   |
   = help: Rename `self` to `cls`
```

Which is slightly... annoying. We know we can remove this from config and add a class method decorator, but this has always felt a bit verbose to use.

Is there a better way to handle this case? is this something ruff would consider fixing?

**Comment by zanieb** [comment]
How would you imagine us fixing this? Taking the full invocation in the config?

```toml
classmethod-decorators = ["classmethod", "pydantic.model_validator(mode=\"before\")", "pydantic.field_validator"]
```

**Comment by dhruvmanila** [comment]
FWIW, the Pydantic docs do recommend to add the `@classmethod` decorator:

> The first argument should be `cls` (and we also recommend you use `@classmethod` below `@model_validator` for proper type checking), [...]
>
> https://docs.pydantic.dev/latest/concepts/validators/#field-validators

And, similarly for field validators:

> A few things to note on validators:
>
> 1. `@field_validators` are "class methods", so the first argument value they receive is the `UserModel` class, not an instance of `UserModel`. We recommend you use the `@classmethod` decorator on them below the `@field_validator` decorator to get proper type checking.
>
> https://docs.pydantic.dev/latest/concepts/validators/#field-validators

Do you use any type checker like mypy or Pyright? Do they work properly?
